diff -r b5e404b928ea src/video/x11/SDL_x11xinput2.c
--- a/src/video/x11/SDL_x11xinput2.c	Wed Oct 11 02:33:55 2017 -0400
+++ b/src/video/x11/SDL_x11xinput2.c	Wed Oct 11 11:29:32 2017 -0600
@@ -73,6 +73,24 @@
 {
     return ( version >= ((wantmajor * 1000) + wantminor) );
 }
+
+static void
+xinput2_normalize_touch_coordinates(SDL_VideoData *videodata, Window window,
+    double in_x, double in_y, float *out_x, float *out_y)
+{
+   int i;
+   for (i = 0; i < videodata->numwindows; i++) {
+      SDL_WindowData *d = videodata->windowlist[i];
+      if (d->xwindow == window) {
+         *out_x = in_x / (d->window->w-1);
+         *out_y = in_y / (d->window->h-1);
+         return;
+      }
+   }
+   // couldn't find the window...
+   *out_x = in_x;
+   *out_y = in_y;
+}
 #endif /* SDL_VIDEO_DRIVER_X11_XINPUT2 */
 
 void
@@ -171,22 +189,31 @@
 #if SDL_VIDEO_DRIVER_X11_XINPUT2_SUPPORTS_MULTITOUCH
         case XI_TouchBegin: {
             const XIDeviceEvent *xev = (const XIDeviceEvent *) cookie->data;
+            float x, y;
+            xinput2_normalize_touch_coordinates(videodata, xev->event,
+                                  xev->event_x, xev->event_y, &x, &y);
             SDL_SendTouch(xev->sourceid,xev->detail,
-                      SDL_TRUE, xev->event_x, xev->event_y, 1.0);
+                      SDL_TRUE, x, y, 1.0);
             return 1;
             }
             break;
         case XI_TouchEnd: {
             const XIDeviceEvent *xev = (const XIDeviceEvent *) cookie->data;
+            float x, y;
+            xinput2_normalize_touch_coordinates(videodata, xev->event,
+                                  xev->event_x, xev->event_y, &x, &y);
             SDL_SendTouch(xev->sourceid,xev->detail,
-                      SDL_FALSE, xev->event_x, xev->event_y, 1.0);
+                      SDL_FALSE, x, y, 1.0);
             return 1;
             }
             break;
         case XI_TouchUpdate: {
             const XIDeviceEvent *xev = (const XIDeviceEvent *) cookie->data;
+            float x, y;
+            xinput2_normalize_touch_coordinates(videodata, xev->event,
+                                  xev->event_x, xev->event_y, &x, &y);
             SDL_SendTouchMotion(xev->sourceid,xev->detail,
-                                xev->event_x, xev->event_y, 1.0);
+                                x, y, 1.0);
             return 1;
             }
             break;
